src/macaron/slsa_analyzer/package_registry/pypi_registry.py

```diff
@@ -0,0 +1,198 @@
+# Copyright (c) 2023 - 2024, Oracle and/or its affiliates. All rights reserved.
+# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/.
+
+"""The module provides abstractions for the pypi package registry."""
+
+import logging
+import os
+import re
+from datetime import datetime
+from typing import Any
+
+from bs4 import BeautifulSoup
+
+from macaron.slsa_analyzer.package_registry.package_registry import PackageRegistry
+from macaron.util import send_get_http, send_get_http_raw
+
+logger: logging.Logger = logging.getLogger(__name__)
+
+
+class PyPIRegistry(PackageRegistry):
+    """This class implements the npm package registry.
+
+    There is no complete and up-to-date API documentation for the npm registry and the endpoints
+    are discovered by manual inspection of links on https://www.npmjs.com.
+    """
+
+    def __init__(
+        self,
+        hostname: str | None = None,
+        attestation_endpoint: str | None = None,
+        request_timeout: int | None = None,
+        enabled: bool = True,
+    ) -> None:
+        """
+        Initialize the npm Registry instance.
+
+        Parameters
+        ----------
+        hostname : str | None
+            The hostname of the npm registry.
+        attestation_endpoint : str | None
+            The attestation REST API.
+        request_timeout : int | None
+            The timeout (in seconds) for requests made to the package registry.
+        enabled: bool
+            Shows whether making REST API calls to npm registry is enabled.
+        """
+        self.hostname = hostname or ""
+        self.attestation_endpoint = attestation_endpoint or ""
+        self.request_timeout = request_timeout or 10
+        self.enabled = enabled
+        super().__init__("pypi Registry")
+
+
+class PyPIApiClient:
+    """PyPI's official and customed API client."""
+
+    def __init__(self, package: str) -> None:
+        self.headers = {"Host": "pypi.org", "Accept": "application/json"}
```

Comment:
I wonder if those headers are strictly necessary for us to use the PyPI API?

Reply:
Fixed this in f900dfef1bae0413a4474f9e8e9101db3c625af1